### PR TITLE
[aievec] to-llvm flow for i8xi8, i16xi16, bf16xbf16 elementwise multiplication

### PR DESCRIPTION
**Pull request by jamestcl-amd**
This PR includes the e2e tests for i8xi8, i16xi16, bf16xbf16 elementwise multiplication going through the to-llvm flow. 

Changes:
- Add mul/srs/broadcast/concat/ext/undef intrinsics to XLLVM. The op names would need a better naming convention since more intrinsics are going to be added to the XLLVM.
- Add aievec-to-llvm conversion pattern for the new intrinsics.
- Add a `FoldAIECastOps` pattern in `AIEVecToLLVM` pass. This pattern folds all the `aievec.cast` ops that are introduced in the vector->aievec pass. Taking this shortcut, we do not need to conditionally introduce `aievec.cast` ops in the vector->aievec conversion patterns. 
- Add op folder to `aievec.cast` op, and remove the `FoldAIECastOps` pattern originally introduced in the `AIEVecOptimizations` pass.
- Add i8xi8_mul_elem, i8xi8_mul_elem_2, i16xi16_mul_elem, i16xi16_mul_elem_2, bf16xbf16_mul_elem, bf16xbf16_mul_elem_2 e2e tests for the to-llvm flow. This includes updating the testbench.cc and the test script. These tests, like other to-cpp tests, go through the simulator to verify the numeric correctness. 
- Add aievec-to-llvm conversion tests for all the new XLLVM ops.
- Add target llvm translation tests.

### CODE REVIEW

File: include/aie/Dialect/XLLVM/IR/XLLVMAIE2IntrOps.td [hunk 1]
```diff
@@ -63,4 +63,18 @@ def VectorSetI512I256IntrOp :
...
+                   VectorOfLengthAndType<[16], [I32]>:$rhs,
+                   I32:$conf)>;
+
+def I512V32Acc32SrsIntrOp :
+    AIEVec2_IntrOp<"I512.v32.acc32.srs",
+        [TypeIs<"res", VectorOfLengthAndType<[32], [I16]>>]>,
+    Arguments<(ins VectorOfLengthAndType<[16], [I64]>:$lhs,
+                   I32:$shft,
+                   I32:$sign)>;
+
```
Comment (jsetoain) [inline include/aie/Dialect/XLLVM/IR/XLLVMAIE2IntrOps.td:79]: You may want to add translation tests for these as well.

**Review by jsetoain — COMMENTED** [review]
Fantastic job so far, James. Thank you! 🙂

**Comment by jamestcl-amd** [comment]
@david-vc @jsetoain this PR is ready for review. Thanks.

**Review by jsetoain — COMMENTED** [review]
In the future, you may want to split these into a bunch of separate patches, one per added intrinsic, for instance. Quicker turn-around, and it's easier to review.

It looks good to me, a lot of great work. Thanks, James!

File: lib/Conversion/AIEVecToLLVM/AIEVecToLLVM.cpp [hunk 2]
```diff
@@ -354,6 +354,110 @@ class MulOpConversion : public mlir::ConvertOpToLLVMPattern<aievec::MulOp> {
...
+        return {DecodedMulElemOp::Kind::BF16_BF16_FP32_16x1x2x1,
+                aiev2_mul_mac_compute_control(0, 0, 2, 3, 1, 0, 0, 0, 0, 0, 0)};
+      }
+    }
+
+    return {DecodedMulElemOp::Kind::UNSUPPORTED, -1};
+  }
+
+  LogicalResult
+  matchAndRewrite(aievec::MulElemOp op, OpAdaptor adaptor,
```
Comment (david-vc) [inline lib/Conversion/AIEVecToLLVM/AIEVecToLLVM.cpp:416]: @jamestcl-amd , @jsetoain : This function is not doing any checking on the number of vector lanes or alternatively the total number of bits of the operands or result. Looking at https://github.com/Xilinx/mlir-aie/blob/11822b078459a6a3ead7d6d3e06ff20c1cb74172/lib/Dialect/AIEVec/IR/AIEVecOps.cpp#L903 I see that this checks consistency, but not the actual number of lanes. So, which piece of are we relying on checking that the number of lanes is correct?

**Review by david-vc — COMMENTED** [review]
LGTM

File: lib/Dialect/AIEVec/IR/AIEVecOps.cpp [hunk 3]
```diff
@@ -205,6 +205,23 @@ ParseResult CastOp::parse(OpAsmParser &parser, OperationState &result) {
...
+// Cast fold method. It will fold with a preceding Cast operation.
+OpFoldResult CastOp::fold(FoldAdaptor adaptor) {
+  auto srcDefOp = getSource().getDefiningOp();
+  if (!srcDefOp)
+    return nullptr;
+
+  auto castOp = dyn_cast<aievec::CastOp>(srcDefOp);
+
+  if (!castOp)
+    return nullptr;
```
[the remark this answers is not in the document]
Comment (jamestcl-amd) [reply lib/Dialect/AIEVec/IR/AIEVecOps.cpp:217]: Will clean up this in the next PR.

File: lib/Dialect/AIEVec/IR/AIEVecOps.cpp [hunk 4]
```diff
@@ -205,6 +205,23 @@ ParseResult CastOp::parse(OpAsmParser &parser, OperationState &result) {
...
+
+  if (castOp.getIsResAcc() == getIsResAcc())
+    return nullptr;
```
[the remark this answers is not in the document]
Comment (jamestcl-amd) [reply lib/Dialect/AIEVec/IR/AIEVecOps.cpp:220]: Nice catch! Will clean up this in the next PR.

File: lib/Conversion/AIEVecToLLVM/AIEVecToLLVM.cpp [hunk 5]
```diff
@@ -517,9 +628,9 @@ class UPDOpConversion : public mlir::ConvertOpToLLVMPattern<aievec::UPDOp> {
...
 
-        // TODO: determine if the undef intrinsic is needed or if an LLVM undef
-        // suffices
-        // destValue = rewriter.create<LLVM::UndefOp>(op->getLoc(), resultType);
+        // TODO: determine if the undef intrinsic is needed or if an LLVM
+        // undef suffices destValue =
+        // rewriter.create<LLVM::UndefOp>(op->getLoc(), resultType);
 
         std::stringstream ss;
         ss << "llvm.aie." << getVectorTypeString(resultType) << ".undef";
```
[the remark this answers is not in the document]
Comment (jamestcl-amd) [reply lib/Conversion/AIEVecToLLVM/AIEVecToLLVM.cpp:636]: Will clean up this in the next PR.

**Comment by jamestcl-amd** [comment]
New updates before merging this PR:
- Separate the to-llvm and to-cpp tests into two mlir test files.
- Add `// REQUIRES: peano` to the to-llvm tests. In this way, CI regression can correctly mark them unsupported (for now). 
- Address review comments on `aievec.cast` folder and update the test accordingly.
- Address review comments to the `testbench.cc`

**Pull request merged**